api/pipeline_spec.proto

```diff
@@ -0,0 +1,332 @@
+
+package kfp.api;
+
+import "google/protobuf/any.proto";
+
+// The spec of a pipeline.
+message PipelineSpec {
+  // The metadata of the pipeline.
+  PipelineInfo pipeline_info = 1;
+
+  // A list of pipeline tasks, which form a DAG.
+  repeated PipelineTaskSpec tasks = 2;
+
+  // The deployment config of the pipeline.
+  // The deployment config can be extended to provide platform specific configs.
+  // The supported config is [PipelineDeploymentConifg]().
+  google.protobuf.Any deployment_config = 3;
+
+  // The version of the sdk, which compiles the spec.
+  string sdk_version = 4;
+
+  // The version of the schema.
+  string schema_version = 5;
+
+  // The definition of the runtime parameter.
+  message RuntimeParameter {
+    // The type of the parameter.
+    enum Type {
+      TYPE_UNSPECIFIED = 0;
+      INT = 1;
+      DOUBLE = 2;
+      STRING = 3;
+    }
+    // Required field. The type of the runtime parameter.
+    Type type = 1;
+    // Optional field. Default value of the runtime parameter. If not set and
+    // the runtime parameter value is not provided during runtime, an error will
+    // be raised.
+    Value default_value = 2;
+  }
+  // The runtime parameters of the pipeline. Keyed by parameter name.
+  map<string, RuntimeParameter> runtime_parameters = 6;
+}
+
+// The spec of a pipeline task.
+message PipelineTaskSpec {
+  // Basic info of a pipeline task.
+  PipelineTaskInfo task_info = 1;
+
+  // The specification of a task input.
+  message InputSpec {
+    // The name of the producer task.
+    string producer_task = 1;
+
+    // The key of the output.
+    string output_key = 2;
+  }
+
+  // Specification for task inputs which connect to the outputs from
+  // upstream tasks. Keyed by input name.
+  map<string, InputSpec> inputs = 2;
+
+  // The specification of a task output.
+  message OutputSpec {
+    // The type of the artifact.
+    PipelineTypeSchema artifact_type = 1;
+
+    // The properties of the artifact.
+    map<string, PipelineValueOrRuntimeParameter> properties = 2;
+
+    // The custom properties of the artifact.
+    map<string, PipelineValueOrRuntimeParameter> custom_properties = 3;
+  }
+
+  // Specification for task outputs. Keyed by output name.
+  map<string, OutputSpec> outputs = 3;
+
+  // Specification for task parameters, which contains the values provided by
+  // the pipeline users or from runtime parameters. Keyed by parameter name.
+  map<string, PipelineValueOrRuntimeParameter> parameters = 4;
+
+  // Label for the executor of the task.
+  // The specification will be specified in the deployment config.
+  // For example:
+  // ```
+  // tasks:
+  // - task_info:
+  //     name: trainer
+  //   executor_label: trainer
+  // deployment_config:
+  //   @type: cloud.ml.pipelines.v1alpha3.proto.PipelineDeploymentConfig
+  //   executors:
+  //     trainer:
+  //       container:
+  //         image: gcr.io/tfx:latest
+  //         args: []
+  // ```
+  string executor_label = 5;
+

```

Comment:
So this sections describes inputs, outputs, executor, upstream_task for PipelinesTaskSpec. What does executor map to? Its more on the lines of a TFX Components, as opposed to a runtime CI/CD engine e.g. Tekton, Argo?

Reply:
executor spec will be specified in deployment config. We split the executor spec from the core IR as the executor spec is usually platform specific. We want to use deployment config extension to decouple the pipeline main structure (which is platform agnostic) and platform specific configs.

Reply:
Yes, we will. The pre-defined executors are described in the PipelineDeploymentConfig in the same doc. Currently, we only added a container spec, which is intended to support a subset of current ContainerOp features.